Bugfixes: Allow any of the deployment types to be used, fix spurious errors on services

luna-duclos opened the pull request:
This PR fixes two issues I encountered while trying out kubeaudit:

1) When using another deployment type than the one specified in types.go, checks would silently fail.
2) When running kubeaudit on a type not known to it (Such as, for example, running `kubeaudit -f <service yaml>`), you'd get an incorrect error that `automountServiceAccountToken: false` needed to be set.

jinankjain commented:
@luna-duclos Thanks for the PR. I had some concerns regarding two new deployments types. Since k8s api wants to deprecate use of v1beta1 and v1beta2 and just use v1 in future (more reference https://kubernetes.io/docs/reference/workloads-18-19/ ). So it would be a good idea to warn users if they are using v1beta1.

jinankjain commented:
@luna-duclos  If you want to get k8s version you can use the same function as one of the subcommands https://github.com/Shopify/kubeaudit/blob/master/cmd/version.go and warning according to version of k8s would be really awesome. :smile:

luna-duclos commented:
I'd prefer perhaps not adding this as part of this PR, It's more complicated than it looks since there might be no cluster setup at all. Would that be ok ?

jinankjain commented on cmd/allowPrivilegeEscalation.go:
```diff
@@ -49,10 +49,12 @@ func checkAllowPrivilegeEscalation(container Container, result *Result) {
-		if len(result.Occurrences) > 0 {
-			results = append(results, result)
-			break
+		if result != nil {
```
Instead of checking `result != nil` it would be better to change the function newResultFromResource to return an err explicit stating that the particular resource type is not supported. This helps in understanding what exactly happened instead of silently ignoring it with `result != nil`

luna-duclos replied:
Would something akin to this work ?

```result, err := newResultFromResource(resource)
		if err == ErrResourceTypeNotSupported {
			continue
		} else if err != nil {
			panic(err)
		}```

luna-duclos commented on cmd/allowPrivilegeEscalation.go:
```diff
@@ -48,10 +48,16 @@ func checkAllowPrivilegeEscalation(container Container, result *Result) {
-		result := newResultFromResource(resource)
-		checkAllowPrivilegeEscalation(container, &result)
+		result, err := newResultFromResource(resource)
+		if err == ErrResourceTypeNotSupported {
```
That code is weird .. ?

If we add new errors they will all be logged as "Resource type not supported" even if they say something else ..
the reason it is checked separately is because we want to skip unknown resource types (ie. if we get a Service, we have no validations to run, so silently skip it)

jinankjain replied:
@luna-duclos I agree if we want to extend that interface further for multiple errors then it should be something like:  
```
switch err {
    case errResourceTypeNotSupported:
}
```

but silently skipping is bit weird because then when we are never telling a user that the audit was skipped for a particular resource. So at least logging that error would be a good idea.

jinankjain replied:
@luna-duclos I did not get it like how exactly you will obtain the resource type which is not supported in `getResources` function ?

jinankjain commented on cmd/errors.go:
```diff
@@ -34,3 +36,7 @@ const (
 )
+
+var (
+	ErrResourceTypeNotSupported = errors.New("Resource type not supported")
```
Perhaps if you want for internal use then it should be errResourceTypeNotSupported

luna-duclos replied:
I removed this var entirely and did things in a better way

luna-duclos commented on cmd/util.go:
```diff
@@ -43,14 +43,31 @@ func isInNamespace(meta metav1.ObjectMeta, namespace string) (valid bool) {
 }
 
-func newResultFromResource(resource k8sRuntime.Object) (result Result) {
+func newResultFromResource(resource k8sRuntime.Object) (*Result, error) {
```
this is no longer applicable either

luna-md commented:
I think the code is much better now, errors are handled generically in each audit, and getResources will only return resources it knows are supported.

I did leave newResult calls into the loop however, since a new result object is needed to be appended into the results slice that is returned.

luna-md commented:
What would be the best way to add tests for this ?

Should I add some a v1beta2 deployments and an extensions/v1beta1 deployment yml to the fixtures and then add tests for those ? Would that be sufficient ?

Perhaps also a Service to validate those are properly skipped ?

klautcomputing commented:
> What would be the best way to add tests for this ?

Create `*v1beta*.yml` fixtures in `fixtures` and take a look at https://github.com/Shopify/kubeaudit/blob/master/cmd/autofix_test.go or https://github.com/Shopify/kubeaudit/blob/master/cmd/cronjob_test.go and do something similar, just so that you can validate that the workload manifests are properly translated.

> Should I add some a v1beta2 deployments and an extensions/v1beta1 deployment yml to the fixtures and then add tests for those ? Would that be sufficient ?

That sounds good to me!

> Perhaps also a Service to validate those are properly skipped ?

If you want to, that'd be great!

luna-duclos commented:
@klautcomputing I changed some of the current fixtures so they vary between all 3 available deployment types, would that be sufficient ?

I also added an unknown_type fixture and validated it was properly skipped when getting manifest files.

klautcomputing commented:
Sorry for not taking care of this way earlier! Thanks again for all your work. 🙂

luna-duclos commented:
Thank you for getting to it! :)
